src/libsystemd/sd-bus/sd-bus.c

```diff
@@ -1644,6 +1644,47 @@ _public_ int sd_bus_get_bus_id(sd_bus *bus, sd_id128_t *id) {
         return 0;
 }
 
+#define COOKIE_CYCLED (UINT32_C(1) << 31)
+
+static uint64_t cookie_inc(uint64_t cookie) {
+
+        /* Stay within the 32bit range, since classic D-Bus can't deal with more */
+        if (cookie >= UINT32_MAX)
+                return COOKIE_CYCLED; /* Don't go back to zero, but use the highest bit for checking
+                                       * whether we are looping. */
+
+        return cookie + 1;
+}
+
+static int next_cookie(sd_bus *b) {
+        uint64_t new_cookie;
+
+        assert(b);
+
+        new_cookie = cookie_inc(b->cookie);
+
+        /* Small optimization: don't bother with checking for cookie reuse until we overran cookiespace at
+         * least once, but then do it thorougly. */
+        if (FLAGS_SET(new_cookie, COOKIE_CYCLED)) {
+                uint32_t i;
+
+                /* Check if the cookie is currently in use. If so, pick the next one */
+                for (i = 0; i < COOKIE_CYCLED; i++) {
+                        if (!ordered_hashmap_contains(b->reply_callbacks, &new_cookie))
+                                goto good;
+
+                        new_cookie = cookie_inc(new_cookie);
+                }
```

Comment:
Hmm, so if get to the `for` loop, `new_cookie` must be `COOKIE_CYCLED`, and it is never increased again. So this is an "infinite" loop, in the sense that we'll always do full 2**31 iterations.

Can we please have a unittest for this?

Reply:
now, it goes like this: 
```
1 → 2 → 3 → … → CC-2 → CC-1 → CC → CC+1 → CC+2 → … → UM-2 → UM-1 → UM
                              ↑                                     ↓
                               ← ← ← ← ← ← ← ← ← ← ← ← ← ← ← ← ← ← ←
```

Were CC is COOKIE_CYCLED is 2^31 and UM is UINT32_MAX-1, i.e. 2^32-1

Reply:
Indeed. I misunderstood the code.

So this looks OK, and I think we can merge it. A test would still be very welcome.
Maybe also add your comment above as a comment in the code.